core/corehttp/gateway_handler.go

```diff
@@ -273,116 +275,84 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
 	http.Redirect(w, r, ipfsPathPrefix+k.String(), http.StatusCreated)
 }
 
-func (i *gatewayHandler) putEmptyDirHandler(w http.ResponseWriter, r *http.Request) {
-	newnode := uio.NewEmptyDirectory()
-
-	key, err := i.node.DAG.Add(newnode)
+func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
+	rootPath, err := path.ParsePath(r.URL.Path)
 	if err != nil {
-		webError(w, "Could not recursively add new node", err, http.StatusInternalServerError)
+		webError(w, "putHandler: ipfs path not valid", err, http.StatusBadRequest)
 		return
 	}
 
-	i.addUserHeaders(w) // ok, _now_ write user's headers.
-	w.Header().Set("IPFS-Hash", key.String())
-	http.Redirect(w, r, ipfsPathPrefix+key.String()+"/", http.StatusCreated)
-}
-
-func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
-	// TODO(cryptix): either ask mildred about the flow of this or rewrite it
-	webErrorWithCode(w, "Sorry, PUT is bugged right now, closing request", errors.New("handler disabled"), http.StatusInternalServerError)
-	return
-	urlPath := r.URL.Path
-	pathext := urlPath[5:]
-	var err error
-	if urlPath == ipfsPathPrefix+"QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn/" {
-		i.putEmptyDirHandler(w, r)
+	rsegs := rootPath.Segments()
+	if rsegs[0] == ipnsPathPrefix {
+		webError(w, "putHandler: updating named entries not supported", errors.New("WritableGateway: ipns put not supported"), http.StatusBadRequest)
 		return
 	}
 
 	var newnode *dag.Node
-	if pathext[len(pathext)-1] == '/' {
+	if rsegs[len(rsegs)-1] == "QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn" {
 		newnode = uio.NewEmptyDirectory()
 	} else {
-		newnode, err = i.newDagFromReader(r.Body)
+		putNode, err := i.newDagFromReader(r.Body)
 		if err != nil {
-			webError(w, "Could not create DAG from request", err, http.StatusInternalServerError)
+			webError(w, "putHandler: Could not create DAG from request", err, http.StatusInternalServerError)
 			return
 		}
+		newnode = putNode
 	}
 
-	ctx, cancel := context.WithCancel(i.node.Context())
-	defer cancel()
-
-	ipfsNode, err := core.Resolve(ctx, i.node, path.Path(urlPath))
-	if err != nil {
-		// FIXME HTTP error code
-		webError(w, "Could not resolve name", err, http.StatusInternalServerError)
-		return
+	var newPath string
+	if len(rsegs) > 1 {
+		newPath = filepath.Join(rsegs[2:]...)
```

Comment:
we dont want to use filepath here, on windows systems this would put forward slashes in the ipfspaths, which i dont think works like we expect it to.